File: matlab/simpleicp.m
function [H, XmovT] = simpleicp(XFix, XMov, nva)

    arguments
        XFix(:,3) {mustBeReal}
        XMov(:,3) {mustBeReal}
        nva.correspondences(1,1) {mustBeInteger, ...
            mustBeGreaterThanOrEqual(nva.correspondences, 10)} = 1000;
        nva.neighbors(1,1) {mustBeInteger, ...
            mustBeGreaterThanOrEqual(nva.neighbors, 2)} = 10;
        nva.minPlanarity(1,1) {mustBeGreaterThanOrEqual(nva.minPlanarity, 0), mustBeReal, ...
            mustBeLessThan(nva.minPlanarity, 1)} = 0.3;
        nva.maxOverlapDistance(1,1) {mustBePositive, mustBeReal} = Inf;
        nva.minChange(1,1) {mustBePositive, mustBeReal} = 3;
        nva.maxIterations(1,1) {mustBePositive, mustBeInteger} = 100;
    end

    tic;
    fprintf('Create point cloud objects ...\n');
    pcFix = pointcloud(XFix(:,1), XFix(:,2), XFix(:,3));
    pcMov = pointcloud(XMov(:,1), XMov(:,2), XMov(:,3));

    if ~isinf(nva.maxOverlapDistance)
        fprintf('Consider partial overlap of point clouds ...\n');
        pcFix.selectInRange([pcMov.x pcMov.y pcMov.z], nva.maxOverlapDistance);
        if numel(pcFix.sel) == 0
            error(['Point clouds do not overlap within maxOverlapDistance = %.5f! ' ...
                   'Consider increasing the value of maxOverlapDistance.'], ...
                nva.maxOverlapDistance);
        end
    end

    fprintf('Select points for correspondences in fixed point cloud ...\n');
    pcFix.selectNPoints(nva.correspondences);
    selOrig = pcFix.sel;

    fprintf('Estimate normals of selected points ...\n');
    pcFix.estimateNormals(nva.neighbors);

    H = eye(4);

    fprintf('Start iterations ...\n');
    for i = 1:nva.maxIterations

        initialDistances = matching(pcFix, pcMov);

        initialDistances = reject(pcFix, pcMov, nva.minPlanarity, initialDistances);

        [dH, residualDistances{i}] = estimateRigidBodyTransformation(...
            pcFix.x(pcFix.sel), pcFix.y(pcFix.sel), pcFix.z(pcFix.sel), ...
            pcFix.nx(pcFix.sel), pcFix.ny(pcFix.sel), pcFix.nz(pcFix.sel), ...
            pcMov.x(pcMov.sel), pcMov.y(pcMov.sel), pcMov.z(pcMov.sel));

        pcMov.transform(dH);

        H = dH*H;
        pcFix.sel = selOrig;

        if i > 1
            if checkConvergenceCriteria(residualDistances{i}, residualDistances{i-1}, nva.minChange)
                fprintf('Convergence criteria fulfilled -> stop iteration!\n');
                break;
            end
        end

        if i == 1
            fprintf('%9s | %15s | %15s | %15s\n', 'Iteration', 'correspondences', ...
                'mean(residuals)', 'std(residuals)');
            fprintf('%9s | %15d | %15.4f | %15.4f\n', 'orig:0', numel(initialDistances), ...
                mean(initialDistances), std(initialDistances));
        end
        fprintf('%9d | %15d | %15.4f | %15.4f\n', i, ...
            numel(residualDistances{i}), mean(residualDistances{i}), std(residualDistances{i}));

    end

    XmovT = [pcMov.x pcMov.y pcMov.z];

    fprintf('Estimated transformation matrix H:\n');
    fprintf('[%12.6f %12.6f %12.6f %12.6f]\n', H(1,1:4));
    fprintf('[%12.6f %12.6f %12.6f %12.6f]\n', H(2,1:4));
    fprintf('[%12.6f %12.6f %12.6f %12.6f]\n', H(3,1:4));
    fprintf('[%12.6f %12.6f %12.6f %12.6f]\n', H(4,1:4));
    fprintf('Finished in %.3f seconds!\n', toc);

end

function distances = matching(pcFix, pcMov)

    pcMov.sel = knnsearch(...
        [pcMov.x pcMov.y pcMov.z], ...
        [pcFix.x(pcFix.sel) pcFix.y(pcFix.sel) pcFix.z(pcFix.sel)]);

    distances = dot(...
        [pcMov.x(pcMov.sel) pcMov.y(pcMov.sel) pcMov.z(pcMov.sel)] - ...
        [pcFix.x(pcFix.sel) pcFix.y(pcFix.sel) pcFix.z(pcFix.sel)], ...
        [pcFix.nx(pcFix.sel) pcFix.ny(pcFix.sel) pcFix.nz(pcFix.sel)], 2);

end

function distances = reject(pcFix, pcMov, minPlanarity, distances)

    planarity = pcFix.planarity(pcFix.sel);

    med = median(distances);
    sigmad = 1.4826 * mad(distances,1);

    idxReject = ...
        distances < (-med-3*sigmad) | ...
        distances > (med+3*sigmad) | ...
        planarity < minPlanarity;

    pcFix.sel(idxReject) = [];
    pcMov.sel(idxReject) = [];
    distances(idxReject) = [];

end

function [H, residuals] = estimateRigidBodyTransformation(...
    xFix, yFix, zFix, ...
    nxFix, nyFix, nzFix, ...
    xMov, yMov, zMov)

    A = [-zMov.*nyFix + yMov.*nzFix ...
          zMov.*nxFix - xMov.*nzFix ...
         -yMov.*nxFix + xMov.*nyFix ...
         nxFix ...
         nyFix ...
         nzFix];

    l = nxFix.*(xFix-xMov) + ...
        nyFix.*(yFix-yMov) + ...
        nzFix.*(zFix-zMov);

    x = A\l;

    residuals = A*x-l;

    R = eulerAnglesToLinearizedRotationMatrix(x(1), x(2), x(3));

    t = x(4:6);

    H = createHomogeneousTransformationMatrix(R, t);

end

function dR = eulerAnglesToLinearizedRotationMatrix(alpha1, alpha2, alpha3)

    dR = [      1 -alpha3  alpha2
           alpha3       1 -alpha1
          -alpha2  alpha1       1];

end

function H = createHomogeneousTransformationMatrix(R, t)

    H = [R          t
         zeros(1,3) 1];

end

function stop = checkConvergenceCriteria(distancesNew, distancesOld, minChange)

    changeOfMean = change(mean(distancesNew), mean(distancesOld));
    changeOfStd = change(std(distancesNew), std(distancesOld));

    if changeOfMean < minChange && changeOfStd < minChange
        stop = true;
    else
        stop = false;
    end

    function p = change(new, old)

        p = abs((new-old)/old*100);

    end

end

File: python/simpleicp.py
"""
Implementation of a rather simple version of the Iterative Closest Point (ICP) algorithm.
"""

import time
from datetime import datetime

import numpy as np
from scipy import spatial, stats

from pointcloud import PointCloud


def log(text):
    """Print text with time stamp."""
    logtime = datetime.now().strftime("%H:%M:%S.%f")[:-3]
    print(f"[{logtime}] {text}")


def matching(pcfix, pcmov):
    """Matching of point clouds."""
    kdtree = spatial.cKDTree(np.column_stack((pcmov.x, pcmov.y, pcmov.z)))
    query_points = np.column_stack(
        (pcfix.x[pcfix.sel], pcfix.y[pcfix.sel], pcfix.z[pcfix.sel])
    )
    _, pcmov.sel = kdtree.query(query_points, k=1, p=2, n_jobs=-1)

    dx = pcmov.x[pcmov.sel] - pcfix.x[pcfix.sel]
    dy = pcmov.y[pcmov.sel] - pcfix.y[pcfix.sel]
    dz = pcmov.z[pcmov.sel] - pcfix.z[pcfix.sel]

    nx = pcfix.nx[pcfix.sel]
    ny = pcfix.ny[pcfix.sel]
    nz = pcfix.nz[pcfix.sel]

    no_correspondences = len(pcfix.sel)
    distances = np.empty(no_correspondences)
    for i in range(0, no_correspondences):
        distances[i] = dx[i] * nx[i] + dy[i] * ny[i] + dz[i] * nz[i]

    return distances


def reject(pcfix, pcmov, min_planarity, distances):
    """Rejection of correspondences on the basis of multiple criterias."""

    planarity = pcfix.planarity[pcfix.sel]

    med = np.median(distances)
    sigmad = stats.median_absolute_deviation(distances)

    keep_distance = [abs(d - med) <= 3 * sigmad for d in distances]
    keep_planarity = [p > min_planarity for p in planarity]

    keep = keep_distance and keep_planarity

    pcfix.sel = pcfix.sel[keep]
    pcmov.sel = pcmov.sel[keep]
    distances = distances[keep]

    return distances


def estimate_rigid_body_transformation(
    x_fix, y_fix, z_fix, nx_fix, ny_fix, nz_fix, x_mov, y_mov, z_mov
):
    """Estimate rigid body transformation for a given set of correspondences."""

    A = np.column_stack(
        (
            -z_mov * ny_fix + y_mov * nz_fix,
            z_mov * nx_fix - x_mov * nz_fix,
            -y_mov * nx_fix + x_mov * ny_fix,
            nx_fix,
            ny_fix,
            nz_fix,
        )
    )

    l = nx_fix * (x_fix - x_mov) + ny_fix * (y_fix - y_mov) + nz_fix * (z_fix - z_mov)

    x, _, _, _ = np.linalg.lstsq(A, l, rcond=None)

    residuals = A @ x - l

    R = euler_angles_to_linearized_rotation_matrix(x[0], x[1], x[2])

    t = x[3:6]

    H = create_homogeneous_transformation_matrix(R, t)

    return H, residuals


def euler_angles_to_linearized_rotation_matrix(alpha1, alpha2, alpha3):
    """Compute linearized rotation matrix from three Euler angles."""

    dR = np.array([[1, -alpha3, alpha2], [alpha3, 1, -alpha1], [-alpha2, alpha1, 1]])

    return dR


def create_homogeneous_transformation_matrix(R, t):
    """Create homogeneous transformation matrix from rotation matrix R and translation vector t."""

    H = np.array(
        [
            [R[0, 0], R[0, 1], R[0, 2], t[0]],
            [R[1, 0], R[1, 1], R[1, 2], t[1]],
            [R[2, 0], R[2, 1], R[2, 2], t[2]],
            [0, 0, 0, 1],
        ]
    )

    return H


def check_convergence_criteria(distances_new, distances_old, min_change):
    """Check if the convergence criteria is met."""
    def change(new, old):
        return np.abs((new - old) / old * 100)

    change_of_mean = change(np.mean(distances_new), np.mean(distances_old))
    change_of_std = change(np.std(distances_new), np.std(distances_old))

    return True if change_of_mean < min_change and change_of_std < min_change else False


def simpleicp(
    X_fix,
    X_mov,
    correspondences=1000,
    neighbors=10,
    min_planarity=0.3,
    min_change=1,
    max_iterations=100,
):
    """Implementation of a rather simple version of the Iterative Closest Point (ICP) algorithm."""

    start_time = time.time()
    log("Create point cloud objects ...")
    pcfix = PointCloud(X_fix[:, 0], X_fix[:, 1], X_fix[:, 2])
    pcmov = PointCloud(X_mov[:, 0], X_mov[:, 1], X_mov[:, 2])

    log("Select points for correspondences in fixed point cloud ...")
    pcfix.select_n_points(correspondences)
    sel_orig = pcfix.sel

    log("Estimate normals of selected points ...")
    pcfix.estimate_normals(neighbors)

    H = np.eye(4)
    residual_distances = []

    log("Start iterations ...")
    for i in range(0, max_iterations):

        initial_distances = matching(pcfix, pcmov)

        # Todo Change initial_distances without return argument
        initial_distances = reject(pcfix, pcmov, min_planarity, initial_distances)

        dH, residuals = estimate_rigid_body_transformation(
            pcfix.x[pcfix.sel],
            pcfix.y[pcfix.sel],
            pcfix.z[pcfix.sel],
            pcfix.nx[pcfix.sel],
            pcfix.ny[pcfix.sel],
            pcfix.nz[pcfix.sel],
            pcmov.x[pcmov.sel],
            pcmov.y[pcmov.sel],
            pcmov.z[pcmov.sel],
        )

        residual_distances.append(residuals)

        pcmov.transform(dH)

        H = dH @ H
        pcfix.sel = sel_orig

        if i > 0:
            if check_convergence_criteria(
                residual_distances[i], residual_distances[i - 1], min_change
            ):
                log("Convergence criteria fulfilled -> stop iteration!")
                break

        if i == 0:
            log(
                f"{'Iteration':9s} | "
                f"{'correspondences':15s} | "
                f"{'mean(residuals)':15s} | "
                f"{'std(residuals)':15s}"
            )
            log(
                f"{0:9d} | "
                f"{len(initial_distances):15d} | "
                f"{np.mean(initial_distances):15.4f} | "
                f"{np.std(initial_distances):15.4f}"
            )
        log(
            f"{i+1:9d} | "
            f"{len(residual_distances[i]):15d} | "
            f"{np.mean(residual_distances[i]):15.4f} | "
            f"{np.std(residual_distances[i]):15.4f}"
        )

    log("Estimated transformation matrix H:")
    log(f"H = [{H[0, 0]:12.6f} {H[0, 1]:12.6f} {H[0, 2]:12.6f} {H[0, 3]:12.6f}]")
    log(f"H = [{H[1, 0]:12.6f} {H[1, 1]:12.6f} {H[1, 2]:12.6f} {H[1, 3]:12.6f}]")
    log(f"H = [{H[2, 0]:12.6f} {H[2, 1]:12.6f} {H[2, 2]:12.6f} {H[2, 3]:12.6f}]")
    log(f"H = [{H[3, 0]:12.6f} {H[3, 1]:12.6f} {H[3, 2]:12.6f} {H[3, 3]:12.6f}]")
    log(f"Finished in {time.time() - start_time:.3f} seconds!")

    return H
